staging/src/k8s.io/client-go/tools/portforward/websocket-connection.go

```diff
@@ -0,0 +1,387 @@
+// stream creation channel is used to communicate dynamically created streams,
+// if the websocket connection detects a stream creation signal.
+func (c *WebsocketConnection) Start(streamCreateCh chan httpstream.Stream, bufferSize int, period time.Duration, deadline time.Duration) {
+	// Initialize and start the ping/pong heartbeat, if necessary. Only client-side connection
+	// needs to run the heartbeat.
+	if period > 0 && deadline > 0 {
+		h := newHeartbeat(c.conn, period, deadline)
+		// Set initial timeout for websocket connection reading.
+		if err := c.conn.SetReadDeadline(time.Now().Add(deadline)); err != nil {
+			klog.Errorf("Websocket initial setting read deadline failed %v", err)
+			return
+		}
+		go h.start()
+	}
+	// Buffer size must correspond to the same size allocated
+	// for the read buffer during websocket client creation. A
+	// difference can cause incomplete connection reads.
+	readBuffer := make([]byte, bufferSize)
+	for {
+		// NextReader() only returns data messages (BinaryMessage or Text
+		// Message). Even though this call will never return control frames
+		// such as ping, pong, or close, this call is necessary for these
+		// message types to be processed. There can only be one reader
+		// at a time, so this reader loop must *not* be run concurrently;
+		// there is no lock for reading. Calling "NextReader()" before the
+		// current reader has been processed will close the current reader.
+		// If the heartbeat read deadline times out, this "NextReader()" will
+		// return an i/o error, and error handling will clean up.
+		messageType, r, err := c.conn.NextReader()
+		if err != nil {
+			websocketErr, ok := err.(*gwebsocket.CloseError)
+			if !ok || websocketErr.Code != gwebsocket.CloseNormalClosure {
+				c.closeAllStreamReaders(fmt.Errorf("next reader: %w", err))
+			}
+			return
+		}
+		// Throttle reading from the websocket connection (note: only one Reader
+		// on the connection can be active at a time).
+		if c.maxBytesPerSec > 0 {
+			r = flowrate.NewReader(r, c.maxBytesPerSec)
+		}
+		// All remote command protocols send/receive only binary data messages.
+		if messageType != gwebsocket.BinaryMessage {
+			c.closeAllStreamReaders(fmt.Errorf("unexpected message type: %d", messageType))
+			return
+		}
+		// Initially, read the websocket stream headers.
+		wsStreamHeaders, err := readWsStreamHeaders(r)
+		if err != nil {
+			c.closeAllStreamReaders(fmt.Errorf("read stream id: %w", err))
+			return
+		}
+		klog.V(5).Infof("websocket stream headers read: %s", wsStreamHeaders.String())
+		streamType := wsStreamHeaders.StreamType
+		streamID := wsStreamHeaders.StreamID
+		// StreamCreate signal means the other websocket connection endpoint
+		// has created a new stream.
+		if streamType == StreamCreate {
+			klog.V(4).Infof("stream create signal: %d", streamID)
+			// Create the stream, but do not send StreamCreate signal.
+			stream, err := c.createStream(wsStreamHeaders.Headers)
```

Comment:
What will limit concurrency of new stream requests?

Reply:
Currently, there is no limit on concurrent requests/streams. I initially implemented a request throttling mechanism with a condition variable on the data structure which stores the streams for each request (`requestStreamsMap`), and this mechanism was validated with a 100k request test using a concurrent request limit of 30. For Alpha I removed it, however, because it was not clear this mechanism would be necessary. And it would probably need significant benchmarking to determine the correct concurrency limit. If we determine we need this mechanism, we can introduce it in Beta.

On a related note, the websocket connection does have flowrate limiting using the same library (`github.com/mxk/go-flowrate/flowrate`) and value as the `UpgradeAwareProxy`.